The append vs overwrite document

pappasilenus opened the pull request:
my magnum opus.  All your comments on the gdocs version have been addressed, I hope.

pappasilenus commented:
Note: you may actually need to pull this and build in sphinx to see the animations for cross and union, since I used HTML markup to scale them and relative links for efficiency.  I've verified it builds correctly in my environment.

gpkc commented:
@pappasilenus Hey, I believe the titles `Overwriting files` and `Appending to files` might be reversed?

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,614 @@
+file-based system that is distributed and supports data of all types
+of files (binary, csv, json, images, etc) from many sources and
+users. That data is processed in parallel across many different jobs
+and pipeline using the Pachyderm Pipeline System (pps). The Pachyderm
```
```suggestion
and pipelines using the Pachyderm Pipeline System (pps). The Pachyderm
```

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,614 @@
+
+### Overwriting files
+
+When putting files into a pfs repo via Pachyderm's `pachctl` utility,
```
Slight tweak, you may want to specify that these are just general api features rather than something specific to `pachctl`.

pappasilenus replied:
Good point, clarified and recommitted!

pappasilenus commented:
> @pappasilenus Hey, I believe the titles `Overwriting files` and `Appending to files` might be reversed?

Good catch, @gpkc !  Fixed, added a transition.  Thanks!

JoeyZwicker commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,616 @@
+### Use cases for large datasets in single files
+
+#### Split and target-file flags
+This default "append" behavior is useful for file types that are often used in data science: line-delimited files and JavaScript Object Notation (json) files. `pachctl` includes the powerful `--split`, `--target-file-bytes` and `--target-file-datums` flags to deal with those files.
```
I think this opening section needs to contextualize the problem better, not just rely on the section title. Why does split even exist? It's also missing a few details. Here's my draft of the wording.

### Splitting large files in Pachyderm

Unlike a system like Git which expects 95+% of your files to be text, Pachyderm does not do intra-file diffing because we work with any file types -- text, JSON, images, binary, etc. Pachyderm diffs content at the per-file level. Therefore, if one bit in content of a file changes, Pachyderm sees that as a new file. Similarly, Pachyderm can only distribute computation at the file-level so if your data is only one large file, it can only be processed by a single worker. 

Because of these reasons, it's pretty common to break up large files into smaller chunks. For simple data types Pachyderm provides the `--split` flag to `put-file` to automatically do this for you. For more complex splitting patterns (e.g. avro or other binary formats), you'll need to manually split your data either at ingest or with a Pachyderm pipeline. 

#### Split and target-file flags
For common file types such as JSON, text or CSV, Pachyderm includes the `--split`, `--target-file-bytes` and `--target-file-datums` flags to deal with those files.

`--split`  will divide those files into chunks based on what a "record" is. In line-delimited files, it's a line.  In json files, it's an object. Pachyderm also support splitting SQL data (specifically pgdump). `--split` takes one argument: `line`, `json`, `sql`.  This argument tells Pachyderm how you want to file delimited. For example, if you use --split `line`, Pachyderm will only divide your file on newline boundaries, never in the middle of a line. Along with the `--split` flag it's common to use additional "target" flags to get better control over the details of the split.

pappasilenus replied:
There's no "sql" option in 1.8.6.  Going to leave that out.

JoeyZwicker commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,616 @@
+COMMIT                           NAME         TYPE COMMITTED          SIZE
+8cce4de3571f46459cbe4d7fe222a466 /my-data.txt dir  About a minute ago 1.071KiB
+```
+Note that the `list-file` command indicates that the line-oriented file we uploaded, "my-data.txt" is actually a directory. That's because the `--split` flag has instructed Pachyderm to split the file up, and it has created a directory with all the chunks in it.  And, as you can see below,  each chunk will be put into a file.  Those are the split-files.  Each split-file will be given a 16-charact ...
````
Add a sentence:

Each split-file will automatically be given a 16-character filename, left-padded with 0. Split does not allow you to define more sophisticated file names right now, but this is a feature we plan to add in the future.

JoeyZwicker commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,616 @@
+
+## Deduplication of output data
+
+Data is not currently deduplicated globally across output repos for performance reasons, only locally, within that repo. If you copy data from an output repo to an input repo, it will be deduplicated globally.
```
Not sure if this is accurate or if we actually support this right now. lets chat with eng.

pappasilenus replied:
I removed this so we can ship this thing, and will add it back if we confirm.

JoeyZwicker commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,616 @@
+
+It's important to remember that what looks like a simple upsert can be a kind of a [fencepost error](https://en.wikipedia.org/wiki/Off-by-one_error#Fencepost_error).  Being "off by one line" in your data can be expensive, consuming processing resources you didn't intend to spend.
+
+### Datums in Pachyderm pipelines
```
Upon reviewing this doc, I think the lifecycle of a datum and everything below here might make sense as a separate doc. Append.overwrite and split is one doc -- then we have the distributed computing doc with is like Datums 101 and this doc is like Datum 201.

pappasilenus replied:
Split and added to this PR.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,614 @@
+- Pipeline definitions use "glob" operators to filter a view of input repositories.
+- Output repositories must merge data from what may be multiple containers running the same pipeline code, at the same time.
+
+When you add in the ability to do "cross" and "union" operators on
```
Links added.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,614 @@
+
+- Files can be put into pfs in "append" or "overwrite" mode.
+- Pipeline definitions use "glob" operators to filter a view of input repositories.
+- Output repositories must merge data from what may be multiple containers running the same pipeline code, at the same time.
```
I'm going to replace "Pipelines" with "Pachyderm Pipeline System" because it's more correct than "pipelines".

pappasilenus commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,614 @@
+little confusing to understand what's actually happening with your
+files!
+
+This document will take you through loading data into pachyderm,
```
Reworked it a bit, but got the gist of the comment.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,614 @@
+COMMIT                           NAME             TYPE COMMITTED    SIZE
+8560235e7d854eae80aa03a33f8927eb /OHVoterData.csv file 1 second ago 153.8KiB
+```
+If you were to re-run the `put-file` command, above, by default, the
````
Added to my internal style guide.  Hard to reconcile my casual, breezy style with jargon. :-)

pappasilenus commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,614 @@
+$ pachctl create-repo voterData
+$ pachctl put-file voterData master -f OHVoterData.csv
+```
+The file will, by default, be placed into the top-level folder of
````
Added to my internal style guide.  Hard to reconcile my casual, breezy style with jargon. :-)

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,396 @@
+
+Unlike a system like Git,
+which expects almost all of your files to be text, 
+Pachyderm does not do intra-file diffing because we work with any file type: 
```
Git doesn't do intra-file diffing either, this is a common misconception, it does more or less the same thing we do, keeps a full copy of each version of the file and content addresses it so that if the same file is stored twice it can reference the same content. The big difference is that git has a lot of tooling around textual diffs such as the `git diff` command, which does textual diffs of files by default, whereas our `diff-file` only does inter-file diffing because it doesn't assume it's text. But in terms of storage our implementation is very similar to git's.

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,396 @@
+it's an object.
+`--split` takes one argument:
+`line`, `json` or [`sql`](../cookbook/splitting.html#pg-dump-sql-support). 
+This argument tell Pachyderm how you want the file split into chunks.
```
```suggestion
This argument tells Pachyderm how you want the file split into chunks.
```

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,396 @@
+
+- `--target-file-datums` will attempt to fill each split-file with the number of datums you specify.  Going back to that same line-delimited 50-line file above, if you use `--split lines --target-file-datums 2`, you'll see the file split into 50 split-files, each of which will have 2 lines.
+
+- Specifying both flags, `--target-file-datums` and `--target-file-bytes`, will result each split-file containing just enough data to satisfy whichever constraint is hit first.  Pachyderm will split the file and then fill the first target split-file with line-based records until it hits the record limit. If it passes the target byte number with just one record, it will move on to the next split-f ...
```
```suggestion
- Specifying both flags, `--target-file-datums` and `--target-file-bytes`, will result in each split-file containing just enough data to satisfy whichever constraint is hit first.  Pachyderm will split the file and then fill the first target split-file with line-based records until it hits the record limit. If it passes the target byte number with just one record, it will move on to the next split-file.  If it hits the target datum number after adding another line, it will move on to the next split-file. Using the example above, if the flags supplied to put-file are `--split lines --target-file-datums 2 --target-file-bytes 100`, it will have the same result as `--target-file-datums 2`, since that's the most compact constraint, and file sizes will hover around 40 bytes.
```

jdoliner replied:
This is just changing "will result each" to "will result in each."

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,396 @@
+count.txt/0000000000000002: Three
+count.txt/0000000000000003: Four
+count.txt/0000000000000004: Five
+count.txt/0000000000000004: Six
```
```suggestion
count.txt/0000000000000005: Six
```

jdoliner commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,396 @@
+count.txt/0000000000000004: Five
+count.txt/0000000000000004: Six
+```
+This would result in one datum being processed in any pipelines that use this repo: the new file `count.txt/0000000000000004`.
````
```suggestion
This would result in one datum being processed in any pipelines that use this repo: the new file `count.txt/0000000000000005`.
```

jdoliner commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,396 @@
+count.txt/0000000000000004: Four
+count.txt/0000000000000005: Five
+``` 
+As far as Pachyderm is concerned, every single file existing has changed, and a new file has been added.  Six datums would be processed by a downstream pipelines.
````
I'd mention explicitly here that the reason for this is that the filename is taken into account when hashing the data for the pipeline. This will only result in 1 new piece of content being stored "Zero", but all of them will get reprocessed.

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,396 @@
+
+The "datum" is the fundamental unit of data processing in Pachyderm pipelines.  It is defined at the file level and filtered by the "globs" you specify in your pipelines.  *What* makes a datum is defined by you.  How do you do that?
+
+When creating a pipeline, you can specify one or more input repos.  Each of these will contain files.  Those files are filtered by the "glob" you specify in the pipeline's defnition, along with the input operators you use.  That determines how the datums you want your pipeline to process appear in the pipeline: globs and input operators, along with other pipeline configuration operators, specify  ...
```
```suggestion
When creating a pipeline, you can specify one or more input repos.  Each of these will contain files.  Those files are filtered by the "glob" you specify in the pipeline's definition, along with the input operators you use.  That determines how the datums you want your pipeline to process appear in the pipeline: globs and input operators, along with other pipeline configuration operators, specify how you would like those datums orchestrated across your processing containers. Pachyderm Pipeline System (pps) processes each datum individually in containers in pods, using Pachyderm File System (pfs) to get the right data to the right code at the right time and merge the results. 
```

jdoliner replied:
This just fixes a small typo, but the diff is big.

jdoliner commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,616 @@
+
+- **repos** contain _files_ in pfs
+- **pipelines** filter and organize those files into _datums_ for processing through _globs_ and _input repo operators_
+- pps will use available resources to process each datum, using pfs to assign datums to containers and merge results in the pipeline's output repo.
```
PPS is the thing that does the assigning, but is reading data from pfs ofc, so I wouldn't say this is really meaningfully inaccurate.

pappasilenus replied:
Making my prose less than meaningfully inaccurate is a low bar, but I'm glad I cleared it. :-)  Seriously, though, the previous draft said the *pipeline* was responsible for this, and I wanted to make sure that folks didn't think their code had to do it.  And it seemed like data orchestration is the job of the PPS to me, not PFS.

jdoliner commented on doc/fundamentals/lifecycle_of_a_datum.md:
````diff
@@ -0,0 +1,292 @@
+
+/pfs/C/my-data-file-3.txt  # from B
+```
+[Cross input](http://docs.pachyderm.io/en/latest/reference/pipeline_spec.html#cross-input) is essentially a cross-product of all the datums, selected by the globs on the repos you're crossing.  It provides a combination of all the datums to the pipeline that uses it as input, treating each combination as a datum.  
````
I think you can drop essentially here, it's very much a cross-product.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,616 @@
+
+When creating a pipeline, you can specify one or more input repos.  Each of these will contain files.  Those files are filtered by the "glob" you specify in the pipeline's defnition, along with the input operators you use.  That determines how the datums you want your pipeline to process appear in the pipeline: globs and input operators, along with other pipeline configuration operators, specify  ...
+
+To summarize:
```
Documentation doesn't follow the "don't repeat yourself" dictum.  It's best to start a section that goes into advanced topics with a short review of what's been covered, and not distract the reader by asking them to "go here to learn this".  It is worthwhile to include a link to the detailed 101 doc if users want to follow it, but it's good to make each one self-contained rather than having folks keep clicking to get what they want.

JoeyZwicker replied:
fair point. I'm ok with that as long as it doesn't go so far as you'll lose the reader with things they already know.

pappasilenus replied:
You always fast forwarded through the "previously on The X Files", didn't you?

pappasilenus commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,616 @@
+- Pachyderm will merge the output of each pod into each output file, with no guarantee of ordering
+- the output files will look as if everything was done in one processing step
+
+If one of your pipelines is written to take all its input, whatever it may be, process it, and put the output into one file, the final output will be one file.  The many datums in your pipeline's input would become one file at the end, with the processing of every datum reflected in that file.
```
Added diagrams

pappasilenus commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,616 @@
+```
+You'd still see the same six datums, in a random order, in your pipeline as it ran though them, but they'd all be in a directory with the same name: C.
+```
+/pfs/C/my-data-file-1.txt  # from A
````
I see by your comment below you understand my intent.  I explained it better and added a diagram.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,616 @@
+      ]
+  }
+```
+In the case of cross inputs, you can't give the repos being crossed identical names, because of that need to avoid name collisions. If each repo had those three files at the top, there would be nine (9) datums overall, which is every permutation of the input files.  You'd see the following datums, in a random order, in your pipeline as it ran though the nine permutations. 
````
I think I made it clearer with the diagram.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,616 @@
+/pfs/A/my-data-file-1.txt
+/pfs/B/my-data-file-3.txt
+```
+Note that you always see both input directories involved in the cross!
````
That's RST.  Since it also handles images better, I decided to convert this file.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,614 @@
+- a set of bits that would be returned when asking for the old branch of OHVoterData.csv & OHVoterData2004.csv and
+- a set of bits that would be appended to that first set to assemble the new, master branch of OHVoterData.csv.
+
+In short, Pachyderm is smart about keeping the minimum set of bits in the object store and assembling the version of the file you (or your code!) have asked for.
```
Done after conversion to rst.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
```diff
@@ -0,0 +1,616 @@
+
+<img alt="Union input animation: shows how each datum is comprised of each datum in every input to the union." src="../_images/union.gif" width="200px" height="200px"/>
+
+For example, let's say you have two input repos, A and B.  Each of them contain three files with the same names: my-data-file-1.txt, my-data-file-2.txt, and my-data-file-3.txt, but different content.  If you were to cross them in a pipeline, the "input" object in the pipeline spec might look like this.
```
This looks like a bug in the syntax highlighter in github.  The directory "_images" starts with a _, which the simplistic highlighter assumes was markup rather than part of the content.  Sigh.  Fixed by the conversion to rst, because it doesn't require me to embed html to resize an image.

pappasilenus commented on doc/fundamentals/append_overwrite.md:
````diff
@@ -0,0 +1,616 @@
+but knew it , almost all men in their degree, some time or other,
+cherish very nearly the same feelings towards the ocean with me.
+```
+What happens to our word files?  Will they get overwritten?  Not as long as each input file--`tts.txt` and `md.txt`--is being treated as a separate datum.  You'll see the data in the "wordcount" repo looking something like this:
````
Good catch.  I clarified this.

JoeyZwicker commented on doc/fundamentals/append_overwrite.rst:
```diff
@@ -0,0 +1,571 @@
+
+``--split`` will divide those files into chunks based on what a “record”
+is. In line-delimited files, it’s a line. In json files, it’s an object.
+``--split`` takes one argument: ``line``, ``json`` or
```
do we support csv as a type here? For some reason I thought we did, but maybe its incorporated within `line`

JoeyZwicker commented on doc/fundamentals/append_overwrite.rst:
```diff
@@ -0,0 +1,571 @@
+
+::
+   
+   could not put file at "/my-data.txt"; a file of type directory is already there
```
to be clear, this isn't anything magical about pachyderm checking split or anything, but simply because the user is trying to create a file with the same name as an existing dir. No changed requested in the doc, but just wanted to clarify.

JoeyZwicker commented on doc/fundamentals/lifecycle_of_a_datum.rst:
```diff
@@ -0,0 +1,549 @@
+is transformed into another datum,
+in this case, another image.
+
+.. image:: 1-1-datum-relationship.gif
```
this is excellent

JoeyZwicker commented on doc/fundamentals/lifecycle_of_a_datum.rst:
```diff
@@ -0,0 +1,549 @@
+    :alt: opencv example animation showing 1 unique datum into 1 unique datum out
+
+
+1:many or 1:N
```
great example

JoeyZwicker commented on doc/fundamentals/lifecycle_of_a_datum.rst:
```diff
@@ -0,0 +1,549 @@
+    :height: 496px
+    :alt: opencv example animation showing 1 unique datum into N unique datums out
+    
+Many:Many or N:M
```
The example is good here. I think the gif can be made a bit clearer. Let's tweak in person this week

JoeyZwicker commented on doc/fundamentals/lifecycle_of_a_datum.rst:
```diff
@@ -0,0 +1,549 @@
+We also highlight the input file at the top of the image,
+so it’s easier to understand when a file from a particular repo is being processed.
+
+.. image:: A-B-union-names.gif
```
the example is good, but the gif is way too confusing. Let's figure out how to simplify. Same for the other union gif.

JoeyZwicker commented:
This was truly a huge project doc so great job!
